aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+            }
+        }
+    }
+}
+
+/// The error type that is used when resolving symbols using the addr2line library
+#[derive(Debug, Error)]
+pub enum Addr2LineError {
+    /// Failed to read the context of the executable or library
+    #[error("Failed to read object data at path {path}")]
+    InvalidObjectPath {
+        /// The object path
+        path: PathBuf,
+        /// The original io::Error
+        source: io::Error,
+    },
+    /// Failed to parse data for an executable or library
+    #[error("Failed to read object data at path {path}")]
+    ParseError {
+        /// The object path
+        path: PathBuf,
+        /// The original error
+        source: addr2line::object::Error,
+    },
+    /// Failed to create the gimli resolver context
+    #[error("Failed to read object data at path {path}")]
+    CreateContextError {
+        /// The object path
+        path: PathBuf,
+        /// The original error
+        source: addr2line::gimli::Error,
+    },
+}
+
+struct Addr2LineObjectContext {
+    ctx: Context<EndianRcSlice<RunTimeEndian>>,
+}
+
+impl Addr2LineObjectContext {
+    pub fn from_object_path(path: &Path) -> Result<Self, Addr2LineError> {
+        let data = std::fs::read(path).map_err(|e| Addr2LineError::InvalidObjectPath {
+            path: path.to_owned(),
+            source: e,
+        })?;
+        let object: addr2line::object::File<_> = addr2line::object::File::parse(&data[..])
+            .map_err(|e| Addr2LineError::ParseError {
+                path: path.to_owned(),
+                source: e,
+            })?;
+        let ctx = Context::new(&object).map_err(|e| Addr2LineError::CreateContextError {
+            path: path.to_owned(),
+            source: e,
+        })?;
+
+        Ok(Self { ctx })
+    }
+}
+
+impl Addr2LineObjectContext {
+    pub fn resolve(&self, address: u64) -> Option<String> {
+        match self.ctx.find_frames(address) {
+            Ok(mut frames) => {
+                let mut result = None;
+
+                while let Ok(Some(frame)) = frames.next() {
+                    let frame = frame.function.and_then(|function_name| {
+                        function_name
+                            .demangle()
+                            .map(|demangled_name| demangled_name.to_string())
+                            .ok()
+                    });
+
+                    // Return the last frame in the stack of frames, since the inlined functions
+                    // are often very unspecific and don't necessarily tell the user
+                    // which function is really executed. E.g.
+                    // - "core::result::Result<T,E>::as_ref"
+                    // - "core::cell::BorrowRefMut::new"
+                    if frame.is_some() {
+                        result = frame;
+                    }
+                }
+
+                result
+            }
+            Err(_) => None,
+        }
+    }
+}
+
+/// Parsed line for /proc/[pid]/maps
```

Comment:
As you've noticed in the other PR we already have proc maps parsing code we use to attach uprobes, and dave just moved it so it can be used to implement USDT probes too.

Reply:
Sure. But that PR isn't merged yet - and it didn't even exist when I started writing this. My understanding is that this can be unified after whatever PR is merged first. Or even after both are merged in the form of a cleanup PR. I can remove the `pub` annotations to avoid this being a breaking API change.

Reply:
We already have some proc maps parsing code here https://github.com/aya-rs/aya/blob/d968094b662be3449624420b76ea2dd239ef657b/aya/src/programs/uprobe.rs#L192. Whatever PR is merged first (I haven't reviewed the USDT PR yet), the code should be shared.

I haven't commented on the details of the API yet, but yes we would definitely not make the proc maps code public, nor the addr2line re-exports or addr2line specific errors. The API should be small and self contained.